### PR TITLE
RangeFinder: refactoring parameter management

### PR DESCRIPTION
**Pull request by 256shadesofgrey**
Created a new class AP_RangeFinder_Params to store the settings there to allow to increase RANGEFINDER_MAX_INSTANCES past 4 (with this commit up to 7 seem to work properly). Created subgroups in RangeFinder accordingly and adjusted constructors and the pointers in the already present drivers.

I added settings for up to 10 range finders to RangeFinder.cpp to allow in future an 8+2 configuration with a range finder for each supported direction. However due to limitations in other parts of the code using more than 7 will result in weird behavior.

As I was working on this I was also wondering, is there any good reason why the number of range finders/proximity sensors is limited to 3 bits? For example [here](https://github.com/ArduPilot/ardupilot/blob/b94e31b21aab369ccb3d6c358780dd3ea4b102ad/libraries/AP_RangeFinder/RangeFinder.h#L175), [here](https://github.com/ArduPilot/ardupilot/blob/b94e31b21aab369ccb3d6c358780dd3ea4b102ad/libraries/AP_Proximity/AP_Proximity.h#L142) and [here](https://github.com/ArduPilot/ardupilot/blob/b94e31b21aab369ccb3d6c358780dd3ea4b102ad/libraries/AP_Proximity/AP_Proximity.h#L143). If anyone here knows, that would greatly help me with future work on this.

The code was tested with 7xVL53L0X (RNGFNDX_TYPE 16).

### CODE REVIEW

**Comment by WickedShell** [comment]
I'm quite excited for the feature. The main thing that this is missing is parameter conversion code. Without it all current setups that have been configured to use range finders will break. This is unfortunately the most tricky part of moving parameters around like this. Would you be willing to look into doing the parameter conversion as well as part of this? (The battery library does provide a valid example of how to manage this).

The other one we should potentially consider doing on this is create a couple more backend specific param tables. (IE `_PIN`, `_SCALING`, `_STOP_PIN`, `_SETTLE` are only needed in Analog and PWM rangefinders, `_FUNCTION`, `_RATIOMETRIC` is purely used in analog rangefinders. `_PWRNRNG` is only used in PWM rangefinders, and `_ADDR` is only used in I2C drivers). The problem with this change is that it makes the conversion code more complicated, but it would reduce the number of parameters sent to the GCS and serves to ensure only the relevant parameters are sent.

**Comment by 256shadesofgrey** [comment]
@WickedShell The problem with adding parameters like `_ADDR` to the backend is that with the current implementation many backends require an instance of `AP_HAL::OwnPtr<AP_HAL::I2CDevice>` to be passed, which in turn requires an address which we don't have until we call `detect()`. To move parameters to the backends we would have to change the design of at least 5 already existing drivers, moving the `hal.i2c_mgr->get_device()` calls inside the drivers and making the `detect()` functions less "intelligent" (and at this point I'm starting to question why we need `detect()` at all).

Everything aside from the first 5 sensors in `RangeFinder::detect_instance()` should be easier to adjust to this method of handling parameters judging by what I see, but I have been wrong before, and I can't test it with anything but VL53L0X.

If I can do this properly for VL53L0X (currently I can't get the `_ADDR` to show in the settings at all after moving it to the driver), I'll apply the same changes to the other range finders and add the commit(s) here. If not, well, this PR is in my opinion already a significant improvement compared to the current master.

I also took a quick look at the parameter conversion. From what I understand this is simply reassigning the index and copying parameters to the new location? If so, and I can just assign arbitrary indexes in the conversion function, I guess I can start parameter indexes at 0 instead of picking holes in the old indexes?

**Comment by WickedShell** [comment]
Sorry for the delayed response, I've had a large pile of stuff slip off my list of e-mails to respond to.

Yeah, I've always wondered why we don't get the I2C device inside the driver. This is a very common pattern in ArduPilot for some reason that I can't explain. (There could be a good technical reason, or it might just be a bad case of copy/paste). That would be a large hold up. I can ask around if there is a technical reason, but thats starting to scope creep really badly on you, which I'd prefer not to do. I've also given the conversion rules a bit more thought, and with the driver specific parameters its getting to be quite a lot of code to carry around. If we could wipe the old rangefinder parameters I think it's fine, but since we can't do that, maybe we should drop the creation of more backend specific params.

So the parameter conversion you need to know the old lookup location from the parameter tree (the primary index + field storage location + data type). The primary index is different between all the vehicles, so you will need to find this for every supported vehicle. Ideally you can use a small table to fill them out (IE that's what the batt monitor does), however because the params were added over time there probably isn't a consistent padding/offset rule for all the old parameters (IE first 3 rangefinders might have the `_ADDR` on sub index `5 + n * 16` but then rangefinder 4 is stored at index 54 for some reason).

You will only need to copy the parameter if it was previously configured, otherwise you can skip copying it. You will also need code to only do the conversion if it hasn't been done already. To do this you look if one of the new parameter storage locations has ever been written to. (And you ensure that this parameter is always saved during the conversion). Again the battery monitor example is doing this case.

**Comment by 256shadesofgrey** [comment]
Ok. So we drop the idea of making some parameters driver specific for now, and I'll just try to figure out how to make the conversion function properly.

Also, I'd like to get an answer to the question in my original comment, since I didn't get an answer from anyone on gitter (which is the reason I ask it here). Is there a technical reason for limiting the number of range finders to 7 (3 bits)? ArduPilot supports 10 directions (and we made a board with 9 sensors), so I would like to remove the 3 bit limitation (possibly in a future PR, or even directly here if removing the `:3` is all it takes in the end). This limitation just seems quite arbitrary to be there for no reason, so I left it alone for now, but it is limiting.

**Comment by 256shadesofgrey** [comment]
@WickedShell Sorry, took a bit longer because of exams. I implemented the conversion function, and with what testing I did it seems to work. However there seem to be conflicts with the current master now. Should I do a rebase onto the new master?

Also, could you please take a look at [this PR](https://github.com/ArduPilot/ardupilot/pull/8893)? It's a small one, and I will need that functionality to add support for [this sensor array](https://github.com/ipmgroup/RangeFinder_9xVL53L0X_hardware).

P.S. Judging by the log of this semaphoreci test the build didn't fail because of my code.
```
Write failed: Broken pipe

fatal: The remote end hung up unexpectedly
```

**Comment by amilcarlucas** [comment]
yes, please rebase it onto master.

**Comment by 256shadesofgrey** [comment]
Ok, done. The conflict was caused by a change in the comments for the PIN parameter.

**Comment by 256shadesofgrey** [comment]
There is a new conflict due to changes in master. Considering it's been almost 2 months since I implemented everything that was requested, I wonder, will this be integrated before there is another conflict if I fix this one now?

**Comment by tridge** [comment]
looks good, just needs a reviewer to double-check the param conversion code in SITL

File: libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+#include "AP_RangeFinder.h"
+
+// table of user settable parameters
+const AP_Param::GroupInfo AP_RangeFinder_Params::var_info[] = {
+    // @Param: TYPE
+    // @DisplayName: Rangefinder type
+    // @Description: What type of rangefinder device that is connected
+	    // @Values: 0:None,1:Analog,2:MaxbotixI2C,3:LidarLiteV2-I2C,5:PX4-PWM,6:BBB-PRU,7:LightWareI2C,8:LightWareSerial,9:Bebop,10:MAVLink,11:uLanding,12:LeddarOne,13:MaxbotixSerial,14:TeraRangerI2C,15:LidarLiteV3-I2C,16:VL53L0X,17:NMEA,18:WASP-LRF,19:BenewakeTF02,20:BenewakeTFmini
+    // @User: Standard
+    AP_GROUPINFO("TYPE",    0, AP_RangeFinder_Params, type, 0),
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp]: please don't use index 0 in new parameter tables (there is an issue with zero)

Comment (256shadesofgrey) [reply libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp]: Fixed.

**Comment by rmackay9** [comment]
I tested the parameter conversion for two existing range finders and bumped into these issues:

- RNGFND_PWRRNG = 13 went missing
- RNGFND_TYPE,18 became 2322
- RNGFND2_TYPE,35 became 6947

I think the _PWRRNG is because there's an extra underscore in the parameter name.

**Comment by rmackay9** [comment]
As a side note there's a little formatting issue in the AP_RangeFinder_Params.cpp's "Values:" line.

**Comment by rmackay9** [comment]
@256shadesofgrey, you're OK to have a look at the parameter upgrade issue I found above?  It's a bit cheeky of me to hassle you considering this PR was submitted 7 months ago .. but.. still.. if you could address those issues I'll try and merge this PR quickly

**Comment by 256shadesofgrey** [comment]
@rmackay9 I'll look into it now.

**Comment by 256shadesofgrey** [comment]
@rmackay9 I couldn't reproduce the issue with RNGFND2_TYPE, it worked for me with both the code you tested with and what I pushed now. But there was indeed an issue with RNGFND(1)_PWRRNG. To solve this I had to change the AP_RangeFinder_PWM driver a bit. RNGFND(1)_TYPE started working when I resolved the PWRRNG issue (I'm guessing that the lack of the PWRRNG parameter caused the parameter conversion to shift some parameters by 1 index, and once it was back, it fell back in place).

**Comment by rmackay9** [comment]
@256shadesofgrey, OK I can test again.  By the way, the way I tested was:

1. loaded current master onto a flight controller
2. reset all parameters to their defaults (MP's full param list has a button for this on the right)
3. change all RNGFND_xxx parameters to new recognisable values (i.e. 1,2,3,4 .. 13)
4. loaded this new code on the board
5. checked the new RNGFND1_xxx parameters were the values from step 3

Txs for making those fixes

**Comment by rmackay9** [comment]
@256shadesofgrey, I've retested and it appears the parameter upgrade doesn't work at all anymore.
Could you try doing the parameter upgrade test yourself?  This might reduce the turn-around time.
256shades-rngfnd-params

**Comment by 256shadesofgrey** [comment]
@rmackay9 Weird, it worked for me, though I only tested it with the parameters you mentioned with the last fix. I'll test/fix it on Monday or Tuesday (don't have the stuff at home).

**Comment by 256shadesofgrey** [comment]
@rmackay9 The conversion function is weird (I just made it analogous to the one in the battery library), it maps old declared indexes to the array positions in the new configuration. So just changing the declared indexes in the new configuration does nothing. That's why there are indexes 23, 49 and 53 in the configuration, but the parameter conversion function uses 0-15.

**Comment by WickedShell** [comment]
As a note if you are repeatedly testing the conversion code you will need to clear the FRAM when you swap back from the converted to the pre converted version, or the conversion can look like it's working if you reuse the same test values. (Just reset the `FORMAT_VERSION` parameter and reboot to clear everything in the FRAM when testing). (Or if testing in SITL you can just delete `eeprom.bin`)

**Comment by rmackay9** [comment]
@256shadesofgrey, so you'll have a peek at the parameter upgrade issue?

**Comment by 256shadesofgrey** [comment]
@rmackay9
Yes, I'll look at it tomorrow, didn't get around to it today.

**Comment by 256shadesofgrey** [comment]
Ok, so I couldn't reproduce it "not working at all", but there was an issue with 
```
RNGFNDX_FUNCTION
RNGFNDX_GNDCLEAR
RNGFNDX_STOP_PIN
```
I tracked the issue down to a problem with string copy when I printed out the parameters during conversion:
```
ConversionInfo[0]: 53 0 1 RNGFND1_TYPE
ConversionInfo[1]: 53 1 1 RNGFND1_PIN
ConversionInfo[2]: 53 2 4 RNGFND1_SCALING
ConversionInfo[3]: 53 3 4 RNGFND1_OFFSET
ConversionInfo[4]: 53 4 1 RNGFND1_FUNCTION�M
ConversionInfo[5]: 53 5 2 RNGFND1_MIN_CM
ConversionInfo[6]: 53 6 2 RNGFND1_MAX_CM
ConversionInfo[7]: 53 7 1 RNGFND1_STOP_PIN�M
ConversionInfo[8]: 53 8 2 RNGFND1_SETTLE
ConversionInfo[9]: 53 9 1 RNGFND1_RMETRIC
ConversionInfo[10]: 53 10 2 RNGFND1_PWRRNG
ConversionInfo[11]: 53 11 1 RNGFND1_GNDCLEAR�M
ConversionInfo[12]: 53 23 1 RNGFND1_ADDR
ConversionInfo[13]: 53 49 5 RNGFND1_POS
ConversionInfo[14]: 53 53 1 RNGFND1_ORIENT
```
As you see there is some garbage at the end of the names of these parameters. This got solved by making sure that the string will be `'\0'`-terminated:
```
char param_name[17] = {0};
```
And just to make absolutely sure after copying:
```
param_name[sizeof(param_name)-1] = '\0';
```
The previous implementation was copied from `AP_BattMonitor`, so it probably suffers from the same bug.

The reason it worked for me originally is probably down to the fact that at the 17th character was accidentally a `'\0'` in that string before.

This is what it looks like for me now:
Master
screenshot from 2019-01-29 18-58-40

What I just pushed:
screenshot from 2019-01-29 19-01-49

**Comment by rmackay9** [comment]
I've re-tested and it seems to work!  I've squashed the commits together here:https://github.com/rmackay9/rmackay9-ardupilot/commits/ipmgroup-rangefinder-7

I'll merge to master once it passes Travis, nicely done!

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request closed without merging**